### DISCUSSION TITLE
Flame specific lint rules

### DISCUSSION THREAD

**Discussion by erickzanardo** [flame-engine/flame#2277] [Ideas]
We could create Flame specific lint rules, to help the community avoid bad practices that are specific to the Flame Engine.

I propose that we use this discussion to talk about possible lint rules, and once we agree on a rule, we create an issue out of it.

Thanks so much to @incendial for reaching out to me and suggesting this idea.

**Comment by erickzanardo** [comment]
I will start by proposing a lint rule to check if the user is instantiating their game directly on the `GameWidget`, inside a build method. Which would cause the game to "reset" everytime Flutter rebuilds the widget where the `GameWidget` is in.

Bad:

```dart
class MyGamePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GameWidget(game: MyFlameGame());
  }
}
```

Good:

```dart
class MyGamePage extends StatefulWidget {
  @override
  State<MyGamePage> createState() => _MyGamePageState();
}

class _MyGamePageState extends State<MyGamePage> {

  late final _game = _MyFlameGame();

  @override
  Widget build(BuildContext context) {
    return GameWidget(game: _game);
  }
}
```

or

```dart
class MyGamePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return GameWidget.controlled(gameFactory: MyFlameGame.new);
  }
}
```

**Reply by incendial** [reply]
Here is the implementation https://github.com/dart-code-checker/dart-code-metrics/pull/1163. Any feedback is welcome!

**Reply by erickzanardo** [reply]
This looks great @incendial !!

**Reply by incendial** [reply]
Thank you ☺️

**Reply by projectitis** [reply]
This is how I am currently doing it. Is it bad?
No state[less] widget involved...

```dart

class MyGame extends FlameGame with SingleGameInstance { ... }

void main() {
  final game = MyGame( ... );

  runApp(GameWidget(game: game));
}
```

**Reply by spydon** [reply]
> This is how I am currently doing it. Is it bad?
> No state[less] widget involved...
> 
> ```dart
> 
> class MyGame extends FlameGame with SingleGameInstance { ... }
> 
> void main() {
>   final game = MyGame( ... );
> 
>   runApp(GameWidget(game: game));
> }
> ```
> 

No, that's fine.

**Comment by wolfenrain** [comment]
A lint rule to check if the `onLoad` needs to be `async` or not. If the body does not use `await` then it shouldn't be `async` so Flame can directly continue in it's loading cycle.

It is just a minor thing but it can optimize code in the long run (especially for the internals of Flame it would help out if we ever make extra decisions based on that)

✅ **Good**:

```dart
class MyComponent extends Component {
  Future<void> onLoad() async {
    await ...
  }
}

class MyComponent extends Component {
  void onLoad() {
    ...
  }
}
```

❌ **Bad**:

```dart
class MyComponent extends Component {
  Future<void> onLoad() async {
    ...
  }
}
```

**Reply by munsterlander** [reply]
Didn't this just change though to a FutureOr https://github.com/flame-engine/flame/pull/2228? So do you still need to check that or should the rule now be to not do the full thing for that method?

**Reply by wolfenrain** [reply]
A `FutureOr` can still become a `Future<void>` or `void`, the type does not change the functionality in this case :P

**Reply by munsterlander** [reply]
Oh for sure. I just was saying if the lint is for async specifically and the new FutureOr allows you to not have to specify it, you may have await in the method without async being declared, right? OR am I misunderstanding how FutureOr can be utilized? I guess my confusion would be, if you leverage FutureOr, do you still always have to declare async or not?

If not, would that break the lint rule?

**Reply by st-pasha** [reply]
There was a discussion about this on dart-lang: https://github.com/dart-lang/linter/issues/1257

**Reply by incendial** [reply]
@wolfenrain Am I right that this one https://dcm.dev/docs/individuals/rules/common/avoid-redundant-async/ won't work, coz you need to additionally change the return type from `Future<void>` to `void`?

**Reply by incendial** [reply]
So, waiting for an update on this one before implementing.

**Reply by spydon** [reply]
> you need to additionally change the return type from Future<void> to void?

That is correct

**Reply by incendial** [reply]
Done https://github.com/dart-code-checker/dart-code-metrics/pull/1168 🎉

**Comment by aytunch** [comment]
I am not sure if this is always a good practice to call super methods in overrides, but if it is then we can add a lint rule.
Currently, if we don't call `super.render(canvas);` there is no lint warning. And we can cover all the overrides which require super calls.
```
@override
  void render(Canvas canvas) {
    // super.render(canvas);
```

**Reply by wolfenrain** [reply]
Methods that require you to call `super.xxx` should have a `@mustCallSuper` which forces that through static analyzing

**Reply by incendial** [reply]
So, no rule is needed, right?

**Reply by spydon** [reply]
Agree, no rule needed for this.

**Comment by pastordougdev** [comment]
We all probably remember this, but I get forgetful sometimes - do not initialize late final variables in the onMount method.

**Reply by incendial** [reply]
@pastordougdev any chance you can share a code sample?

**Reply by spydon** [reply]
```dart
class MyComponent extends Component {
  late final int x;

  @override
  void onMount() {
    x = 1;
  }
}
```

**Reply by incendial** [reply]
Thank you!

**Reply by incendial** [reply]
But initialising any other type of class field is allowed, right?

**Reply by spydon** [reply]
Indeed, as long as they are not `late final` since you can remove a component and add it again and then `onMount` will run again.

**Reply by incendial** [reply]
Here is the implementation https://github.com/dart-code-checker/dart-code-metrics/pull/1165

**Comment by munsterlander** [comment]
Do not create Vector2 in update as well.

**Reply by pastordougdev** [reply]
Is this a performance issue, or does it cause crazy results?  I tend to try not to create Vector2's in update, but just curious.

**Reply by munsterlander** [reply]
Every tick you would be creating a vector2, when you probably just need to update a vector2 (position of a character or such).  Creating then adds overhead and memory when it's not needed - usually.

**Reply by pastordougdev** [reply]
Agreed.  I try to keep object creation down to a minimum in update methods.  Vector2 has a good number of methods and properties!  I know that Vector2 is built upon NotifyingVector so I didn't know if there was something going on behind the scenes that could be creating any additional issues.  :-)

**Reply by spydon** [reply]
It's the other way around, `NotifyingVector` extends `Vector2`.

**Reply by incendial** [reply]
Can anyone provide an example so I can base the tests on it?

**Reply by spydon** [reply]
```dart
class MyComponent extends Component {
  @override
  void update(double dt) {
    final newVector = Vector2(10, 10);
  }
}
```
@incendial just something like that? So here it should report that you shouldn't create new `Vector2` objects within the `update` method.

**Reply by incendial** [reply]
Yes, thank you! Can you also add a "good" example? Or a good one would be just not to create at all?

**Reply by spydon** [reply]
```dart
class MyComponent extends Component {
  final _temporaryVector = Vector2.zero();

  @override
  void update(double dt) {
    _temporaryVector.setValues(10, 10);
  }
}
```
There are a lot of ways to change the values of a vector, but this is one of the basic ones.

Another case where a new `Vector2` is created is from vector operations, like this (but this might be harder to check for?):

```dart
class MyComponent extends Component {
  final vector1 = Vector2(10, 10);
  final vector2 = Vector2(20, 20);

  @override
  void update(double dt) {
    final addVector = vector1 + vector2;
    final subVector = vector1 - vector2;
  }
}
```

And in this case you should instead do this:

```dart
class MyComponent extends Component {
  final vector1 = Vector2(10, 10);
  final vector2 = Vector2(20, 20);

  final _temporaryVector = Vector2.zero();

  @override
  void update(double dt) {
    _temporaryVector
      ..setFrom(vector1)
      ..add(vector2);
  }
}
```

**Reply by incendial** [reply]
Great, thank you again!

**Reply by incendial** [reply]
@spydon for the second example, I can simply ban all binary operations (like +, -, *, /) when both operands are vectors, will this work or there are some correct cases to have them?

**Reply by spydon** [reply]
That should work!

**Reply by incendial** [reply]
Great! And another one https://github.com/dart-code-checker/dart-code-metrics/pull/1166 😁

**Reply by incendial** [reply]
Should Vector3 and Vector4 be linted as well or Flame allows only to work with Vector2?

**Reply by spydon** [reply]
I think only `Vector2` is fine for now, until we go 3D. 😉

**Comment by incendial** [comment]
All four rules are now available in prerelease version 5.6.0-dev.1 https://pub.dev/packages/dart_code_metrics/versions/5.6.0-dev.1/changelog! 🚀

Any feedback is very welcome! ☺️

**Reply by spydon** [reply]
Good job! Will try them out when I get the chance!

**Reply by munsterlander** [reply]
This is really cool. I will add this package today as well!

**Reply by incendial** [reply]
Thank you, if you need anything else to be linted - just let me know 😉

**Comment by grukillerkams** [comment]


**Reply by spydon** [reply]
@jotterkain please don't post random stuff without content on several different posts. If you want to say +1, just do a thumbs up reaction instead. Thanks.
